### DISCUSSION TITLE
Max Length

### DISCUSSION THREAD

**Discussion by TonyGravagno** [araera111/mysql-to-zod#28] [Ideas]
For types that allow a maximum **size** value, I'd like to add that into the Zod schema.

This is not related to variable sizing (5-10 characters) or required sizing (must be 6). The goal for this challenge is only to ensure that a value does not exceed MySQL limits.

```sql
CREATE TABLE `airport` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `long_name` varchar(100) DEFAULT NULL,
  `short_name` varchar(80) DEFAULT NULL,
  `geo_location_id` bigint(20) unsigned DEFAULT NULL,
  `iata` tinytext DEFAULT NULL,
```

```ts
  id: z.number().nonnegative(),
  longName: z.string().max(100).nullable(),
  shortName: z.string().max(80).nullable(),
  geoLocationId: z.number().int().nonnegative().lte(99999999999999999999).nullable(),
  iata: z.string().nullable(),
```

Numeric length can also be validated with a transform to temporarily convert the number to a string, check the max length, and then return it to a number. The implementation is subject to testing.

As discussed in other exchanges, for any enhancements like this, an option can be added to the config.js to allow toggling the feature and perhaps to modify the implementation.

**Comment by araera111** [comment]
I think I can implement this.
We need to be careful about unsigned, signed, nonnegative, etc.
We need to consider how to write the options.

**Reply by TonyGravagno** [reply]
Just so you know, I have been working on this. It is "simple" but there are nuances that must be handled. I will submit the updates to you in a gradual progression so that you can make decisions about what to accept without having to accept or reject everything.

For example, options will certainly be required. The need for specific options will be seen in each new phase that has hardcoded text or where other choices have been made temporarily. So you will see hardcoded choices in one phase (build/branch) and then in the next phase you will see my attempt to implement options for a developer to make their own choices.

If you are tempted to look at my repo, please don't for now. As I write the code and get each piece in place, I realize new details are required, so what is there is subject to change. When ready, I will provide a new branch tree for you to evaluate.

**Comment by TonyGravagno** [comment]
I welcome you to take a look at https://github.com/TonyGravagno/mysql-to-zod/tree/tg-length-2

## Options

New options on the options.schema.zod node:

```ts
zod: {
  implementation: [],
  references: [],
  maxLength: {
    active: true,
    inline: null, // "Max of ${limit} characters.",
    global: "Too many characters. Maximum ${limit}."
  }
},
```

- maxLength and its suboptions are all optional
- if active is false no special code is produced.
- if schema.inline is true, no special code is produced - this functionality relies on a call to a new function in globalSchema.
- inline and global are nullable strings for error messages when a field length is greater than the table contraint.
  - if inline is not null, the text is used as an error message in all schema that has a max length constraint.
  - if global is not null, the text is used as an error message for schema that does not have its own inline error message.
- the `${limit}` placeholder is a common substitution string which can be present anywhere in the message.
  - The variable here must be "limit".
  - The placeholder is a true substitute in the global error message, the actual value is substituted.
  - The placeholder is hard-substituted into the inline schema: "Max is 20 characters" This is because the inline refinement with a variable would make the code too bulky. This can be enhanced later if desired.

## Output

### Normal/Pre-Mod
#### schema.inline=true -or- zod.maxLength.active=false
#### No change to mysql-to-zod schema

```ts
export const PersonSchema = z.object({
  id: z.number(),
  first_name: z.string().nullish(),
  last_name: z.string(),
  region_id: z.number().nullish(),
  previous_names: z.string().nullish(),
  birth_date: z.date().nullish(),
```

### Normal/Pre-Mod: inline=false, zod.maxLength.active=false
#### No change to mysql-to-zod schema
#### Note the diverse SQL types and nullability

<details>
<summary>schema.ts</summary>

```ts
export const PersonSchema = z.object({
  id: globalSchema.mysqlBIGINT,
  first_name: globalSchema.mysqlVARCHAR.nullish(),
  last_name: globalSchema.mysqlVARCHAR,
  region_id: globalSchema.mysqlBIGINT.nullish(),
  previous_names: globalSchema.mysqlLONGTEXT.nullish(),
  birth_date: globalSchema.mysqlDATE.nullish(),
```

</details>

### New
#### zod.maxLength.active=true
#### zod.maxLength.inline=null
#### zod.maxLength.global=null

<details>
<summary>schema.ts</summary>

```ts
export const PersonSchema = z.object({
  id: globalSchema.mysqlBIGINT.refine((arg) => globalSchema.maxLength(arg, 20)), // length 20, calls maxLength refinement
  first_name: globalSchema.mysqlVARCHAR
    .refine((arg) => globalSchema.maxLength(arg, 60))
    .nullish(), // nullish/nullable and other validation follow the refinement
  last_name: globalSchema.mysqlVARCHAR.refine((arg) =>
    globalSchema.maxLength(arg, 60),
  ),
  legal_name: globalSchema.mysqlVARCHAR
    .refine((arg) => globalSchema.maxLength(arg, 60))
    .nullish(),
  nick_name: globalSchema.mysqlVARCHAR
    .refine((arg) => globalSchema.maxLength(arg, 60))
    .nullish(),
  region_id: globalSchema.mysqlBIGINT
    .refine((arg) => globalSchema.maxLength(arg, 20)) // 20 digits on an integer seems like a constraint issue
    .nullish(),
  disambiguation: globalSchema.mysqlVARCHAR
    .refine((arg) => globalSchema.maxLength(arg, 200)) // large limit
    .nullish(),
  previous_names: globalSchema.mysqlLONGTEXT.nullish(), // no limit on this LongText field
  birth_date: globalSchema.mysqlDATE.nullish(), // no limit on Date field even when DB has a constraint (notes later)
```

</details>

<details>
<summary>globalschema.ts</summary>

```ts
import { z, RefinementCtx } from "zod";
export const globalSchema = {
  mysqlBIGINT: z.number(),
  mysqlVARCHAR: z.string(),
  mysqlLONGTEXT: z.string(),
  mysqlDATE: z.date(),
  mysqlTIMESTAMP: z.date(),
  mysqlINT: z.number(),

  maxLength: (arg: any, limit: number, ctx?: RefinementCtx): boolean => { // function is in globalSchema for easy access
    if (arg?.toString()?.length > limit) {
      if (ctx)
        ctx.addIssue({
          code: z.ZodIssueCode.too_big,
          maximum: limit,
          type: typeof arg === "number" ? "number" : "string",
          inclusive: true, // no message, default from Zod too_big code is used
        });
      return false;
    }
    return true;
  },
};
```

</details>

### Now using options to specify error message text

```ts
  maxLength: {
    active: true,
    inline: "Max of ${limit} characters.",
    global: "Too many characters. Maximum ${limit}."
  }
```

#### globalSchema.ts (only code that changed)

```ts
        ctx.addIssue({
          code: z.ZodIssueCode.too_big,
          maximum: limit,
          type: typeof arg === "number" ? "number" : "string",
          inclusive: true,
          message: `Too many characters. Maximum ${limit}.`,
        });
```

<details>
<summary>schema.ts</summary>

```ts
export const PersonSchema = z.object({
  id: globalSchema.mysqlBIGINT.refine(
    (arg) => globalSchema.maxLength(arg, 20),
    `Max of 20 characters.`, // message is inline and fixed, not variable
  ),
  first_name: globalSchema.mysqlVARCHAR
    .refine((arg) => globalSchema.maxLength(arg, 60), `Max of 60 characters.`) // remove a message here to use global default
    .nullish(),
  last_name: globalSchema.mysqlVARCHAR.refine(
    (arg) => globalSchema.maxLength(arg, 60),
    `Max of 60 characters.`,
  ),
  legal_name: globalSchema.mysqlVARCHAR
    .refine((arg) => globalSchema.maxLength(arg, 60), `Max of 60 characters.`)
    .nullish(),
  nick_name: globalSchema.mysqlVARCHAR
    .refine((arg) => globalSchema.maxLength(arg, 60), `Max of 60 characters.`)
    .nullish(),
  region_id: globalSchema.mysqlBIGINT
    .refine((arg) => globalSchema.maxLength(arg, 20), `Max of 20 characters.`)
    .nullish(),
  disambiguation: globalSchema.mysqlVARCHAR
    .refine((arg) => globalSchema.maxLength(arg, 200), `Max of 200 characters.`)
    .nullish(),
  previous_names: globalSchema.mysqlLONGTEXT.nullish(), // still no max on LongText
  birth_date: globalSchema.mysqlDATE.nullish(), // or Date
```

</details>

## Notes

- Documentation hasn't been written yet (except the above).
- Types DATE and TIMESTAMP are explicitly excluded from length check.
  - The TIME types report a length parameter but this is actually the "fsp" (fractional seconds precision).
  - YEAR has a parameter reported as length, but this has been deprecated and shouldn't be supported for new apps.
  - [Ref](https://dev.mysql.com/doc/refman/8.4/en/date-and-time-type-syntax.html) and other pages on these types.
- If we support Labels, maybe from column-level comments, error messages on each field can include the name of the field : "Name field has a limit of 20 characters" ... "City field has a limit of 30 characters".
- The Zod schema function to do this is `refine()`. It's a little verbose. Zod can be extended later to make this more concise.
- The Zod schema `superRefine` can be used with the same maxLength function. This is why the `ctx` option is there.
- The feature here is explicitly "max length". This is different from numeric "max value". A constraint of 2 on a numeric gives it a range from "-9" to "99" ... only two characters. Compare this to a numeric range, which uses the `.lte` function for Less Than or Equal.
  - Yes, there is a constraint for max number of digits on a field. [Ref](https://dev.mysql.com/doc/refman/8.4/en/numeric-type-syntax.html).
- Similarly, the Boolean/Bool type is a synonym for TinyInt(1)
- I will look more closely at other details and anomalies to ensure values incorrectly reported as "length" are not processed by maxLength.
- This feature should be considered alpha for now, not to be merged into core/production. It needs to be tested with real data, and more examples of contraints that identify as "length".
 - It might be helpful (optionally) to provide a default length constraint depending on the data type. For example, TINYINT is defined as a 1-Byte storage field with a max range of -128 to 127 (255 unsigned). Knowing these details we could have an option to constrain entry fields to 3 or 4 characters, and/or to add data value limits on the schema as well.
 - Developers should be aware of overflow handling in MySQL and other databases. They may want to leave some processing to the DBMS, or they/we may want to define a library of custom schema constraints to handle specific server nuances. We can't do too much fine tuning on this for now. [Ref](https://dev.mysql.com/doc/refman/8.4/en/out-of-range-and-overflow.html)

#### That's it for now. 😁

**Reply by TonyGravagno** [reply]
The RefinementCtx is a persistent context which accumulates a collection of errors.  
To use this in the current code:

- Note that there is already a context field on `maxLength( arg, limit, ctx )`.
- Manually change the generated Zod schema as follows:

```ts
  legal_name: globalSchema.mysqlVARCHAR
    .superRefine((arg,ctx) => globalSchema.maxLength(arg, 60, ctx))
    .nullish(),
```

A 'useContext' option can be added for all schema to use this as the default.

It might seem redundant to set the option zod.maxLength.inline value to a text message and to use ctx. It's not. The choice is up to the developer. Use the messages inline for some fields and for others use the message in context (global text value).

**Comment by TonyGravagno** [comment]
Please comment when you have time. Thank you!

**Comment by araera111** [comment]
Thank you very much for your significant contribution.

I find this feature quite interesting and am considering using it in practice.

I have merged the changes into the develop branch and tested to ensure that existing tests are not broken. Specifically, I modified the import statement for the composeGlobalSchema function to include RefinementCtx only when option?.schema?.inline === false and option?.schema?.zod?.maxLength?.active === true. Additionally, I have addressed any issues where existing tests were failing.

I have pushed the latest changes to the develop branch. It would be great if you could verify that everything is functioning correctly.

I believe it would be appropriate to merge this feature into main, but it might also be beneficial to include an experimental flag to alert users to its experimental nature.

Thank you for your attention to this matter.

**Reply by TonyGravagno** [reply]
I'm pulling from https://github.com/araera111/mysql-to-zod/tree/develop now to continue testing.

Thank you VERY VERY much!! :)

**Comment by TonyGravagno** [comment]
Some things are not working in this latest Dev version:

1. The fix to use the options.tables parameter is not in this, so all tables are processed. That's OK, this can be merged in later.

2. In this line, the readFileSync fails on the first run because there is no globalSchema.ts. https://github.com/araera111/mysql-to-zod/blob/376e5c39a75f960a401edad6f0bfc2d45481722b/src/process/outputToFile/mergeGlobalConfig.ts#L37
I hacked a new "exists" function because the node:fs.exists function is deprecated:

```ts
const exists = (path: string): boolean => {
	try {
		const fd = openSync(path, 'r');
		closeSync(fd);
		return true;
	} catch (err) {
		return false;
	}
};
```

So now the readFileSync looks like this:

```ts
	let oldAst = toKeyValuePair('');
	if (exists(path)) {
		const oldGlobalSchemaText = readFileSync(path, "utf-8");
		oldAst = toKeyValuePair(oldGlobalSchemaText);
	}
	const newAst = toKeyValuePair(globalSchema);
	const res = [...oldAst, ...newAst];
```

Unless there is a better way to do that, perhaps that `exists` function can be added as a global export.

3. Now, in `const newAst = toKeyValuePair(globalSchema);`, the toKeyValuePair function parses on colons ( ":" ). The schema now has an embedded function, which includes the colon character, so that function breaks up the new maxLength function. 


If you merge in your fix for tables, I can fix everything else. ... Or I will be happy to do all of it. 😁 I just want to give you an opportunity first to make whatever decisions you wish with your software - and I will comply with whatever decisions you make.

Thanks!

**Comment by araera111** [comment]
Thank you for your message.

For the first issue, please pull the latest develop branch, and it should be resolved. If the filterTable function is being used in getTables.ts, only the specified tables should be processed as expected.

Regarding the second point, I’ve addressed this by switching to the existsSync function, which is not deprecated, unlike exists. The code now checks for the existence of the globalSchema before attempting to merge it. If the schema doesn’t exist, only the newly generated one will be processed. This fix is also available in the latest develop branch, so feel free to pull it.

As for the third issue, I agree that this behavior can occur. I’ll work on adjusting the function. Could you kindly provide a minimal version of your globalSchema where this issue arises? I’ll modify the function to ensure the old and new schemas can be merged correctly. Currently, it's a simple key-value parsing process, but if that approach doesn’t work, I might consider parsing via AST or possibly using peggy. Either way, sharing a case where it breaks would be very helpful.

Thank you again for your input, and I look forward to your feedback.

**Reply by TonyGravagno** [reply]
This is what globalSchema.ts should look like:

```ts
import { z, RefinementCtx } from "zod";
export const globalSchema = {
  mysqlBIGINT: z.number(),
  mysqlVARCHAR: z.string(),
  mysqlLONGTEXT: z.string(),
  mysqlDATE: z.date(),
  mysqlTIMESTAMP: z.date(),
  mysqlINT: z.number(),
  mysqlTINYINT: z.number(),
  mysqlSMALLINT: z.number(),
  mysqlMEDIUMINT: z.number(),
  mysqlDECIMAL: z.string(),
  mysqlFLOAT: z.number(),
  mysqlDOUBLE: z.number(),
  mysqlTIME: z.string(),
  mysqlYEAR: z.number(),
  mysqlDATETIME: z.date(),
  maxLength: (arg: any, limit: number, ctx?: RefinementCtx) : boolean => {
    if (arg?.toString()?.length > limit) {
      if (ctx)
        ctx.addIssue({
          code: z.ZodIssueCode.too_big,
          maximum: limit,
          type: typeof arg === "number" ? "number" : "string",
          inclusive: true,
     message: `Too many characters. Maximum ${limit}.`,
       });
      return false;
    }
    return true;
  },
}
```

This is what it looks like when the maxLength parameter is split on colons:

```ts
import { z, RefinementCtx } from "zod";
export const globalSchema = {
  mysqlBIGINT: z.number(),
  mysqlVARCHAR: z.string(),
  mysqlLONGTEXT: z.string(),
  mysqlDATE: z.date(),
  mysqlTIMESTAMP: z.date(),
  mysqlINT: z.number(),
  mysqlTINYINT: z.number(),
  mysqlSMALLINT: z.number(),
  mysqlMEDIUMINT: z.number(),
  mysqlDECIMAL: z.string(),
  mysqlFLOAT: z.number(),
  mysqlDOUBLE: z.number(),
  mysqlTIME: z.string(),
  mysqlYEAR: z.number(),
  mysqlDATETIME: z.date(),
  maxLength: (arg
  code: z.ZodIssueCode.too_big,
  maximum: limit,
  type: typeof arg === "number" ? "number"
  inclusive: true,
  message: `Too many characters. Maximum ${limit}.`,};
```

I think this can be addressed cleanly as follows:
1) In composeGlobalSchema.ts we collapse functions like maxLengthFunction into a single line, replacing "\n" with "~EOL~" (EndOfLine).
1a) If a function doesn't have a EOL we add one at the end anyway. 
2) When that gets into mergeGlobal.ts and toKeyValuePair(), the flatMap can check for the placeholder text.
3) If 'x' includes "~EOL~", we don't split to get the key+value, we slice it to extract the function name as key and the function body as value.
4) After the globalSchema is processed, do a final replace of the placeholder with "\n":
`const fullGlobalTs = \`${importZod}${exportGlobal}${body}${end}\`.replace("~EOL~","\n");`
4a) If an extra EOL was added in 1a, remove it : `.replace("\n\n","\n")`

Steps 1 and 4 can be encapsulated into functions : collapseFunctions() and expandFunctions()

**Comment by TonyGravagno** [comment]
I will pull from Develop again after you have reviewed this. For now I'm getting back on app code.

**Comment by araera111** [comment]
Thank you very much for your cooperation. I have implemented support for merging across multiple lines, taking inspiration from your suggested approach. I also created tests for the cases you provided, and they have passed successfully. Additionally, I realized I had overlooked the merging of import statements, so I have implemented that as well.

If you could merge the develop branch and give it a try, I would greatly appreciate it.

**Comment by TonyGravagno** [comment]
As noted in #170 there is still an issue with the develop branch but I'm sure it will be resolved soon.  We're making a lot of changes now and I think it would be prudent to continue adding to develop, and testing there until we are very comfortable with it. To help with testing, I have created another branch off of develop for my own development of numeric constraints for #54 and some other enhancements.

To be clear, I do not intend for the current implementation of this MaxLength functionality to be the final version. This has been implemented in a general way and needs to be refined. This first implementation is a only base, a starting point, for types that truly support field size/length. This base can be used for minLength as well for some string types.

Rather than calling to a maxLength or minLength function, some Zod types can simply use the type's `.max()` or `.min()` function. However, for some numeric types, `.max()` means "quantity", not "size/length" of the field. This depends on whether the field is an integer type or a float type.

Also, the length specification on a date/time type restricts the stored and displayed precision of seconds, not the size of the entire field.

And for a char() type, the length specification is both a min and a max, absolute specification.

These and other nuances will be addressed in specific tickets, discussions, and updates.
